File: espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/w3c/TickTest.java
package io.appium.espressoserver.test.helpers.w3c;


import org.junit.Test;

import java.util.List;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.CompletionService;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorCompletionService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException;
import io.appium.espressoserver.lib.helpers.w3c.adapter.DummyW3CActionAdapter;
import io.appium.espressoserver.lib.helpers.w3c.adapter.W3CActionAdapter;
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.BaseDispatchResult;
import io.appium.espressoserver.lib.helpers.w3c.models.ActionObject;
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType;
import io.appium.espressoserver.lib.helpers.w3c.models.Origin;
import io.appium.espressoserver.lib.helpers.w3c.models.Tick;
import io.appium.espressoserver.lib.helpers.w3c.state.InputStateTable;
import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState;
import io.appium.espressoserver.lib.helpers.w3c.state.PointerInputState;

import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.KEY_DOWN;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.KEY_UP;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.PAUSE;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.POINTER_DOWN;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.POINTER_MOVE;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.KEY;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.NONE;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.POINTER;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType.TOUCH;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.VIEWPORT;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class TickTest {

    @Test
    public void shouldCalculateMaxDurationZeroIfNoDurations() throws ExecutionException, InterruptedException {
        Tick tick = new Tick();
        tick.addAction(new ActionObject());
        assertEquals(tick.calculateTickDuration(), 0);
    }

    @Test
    public void shouldCalculateMaxDuration() {

        Long[] valueOne =    new Long[] { 10L, 30L, 0L };
        Long[] valueTwo =    new Long[] { 20L, 10L, 1L };
        long[] expectedMax = new long[] { 20L, 30L, 1L };

        for (int i=0; i<valueOne.length; i++) {
            Tick tick = new Tick();
            ActionObject actionObjectOne = new ActionObject();
            actionObjectOne.setType(NONE);
            actionObjectOne.setSubType(PAUSE);
            actionObjectOne.setDuration(valueOne[i]);

            ActionObject actionObjectTwo = new ActionObject();
            actionObjectTwo.setType(POINTER);
            actionObjectTwo.setSubType(POINTER_MOVE);
            actionObjectTwo.setDuration(valueTwo[i]);

            ActionObject actionObjectThree = new ActionObject();
            actionObjectTwo.setType(POINTER);
            actionObjectTwo.setSubType(POINTER_MOVE);

            tick.addAction(actionObjectOne);
            tick.addAction(actionObjectTwo);
            tick.addAction(actionObjectThree);

            assertEquals(tick.calculateTickDuration(), expectedMax[i]);
        }
    }

    @Test
    public void shouldAddKeyInputStateWhenRunningActions() throws AppiumException {
        InputStateTable inputStateTable = new InputStateTable();
        Tick tick = new Tick();
        String sourceId = "something1";
        ActionObject actionObject = new ActionObject(sourceId, KEY, null, 0);
        actionObject.setSubType(KEY_DOWN);
        actionObject.setValue("F");
        tick.addAction(actionObject);
        assertFalse(inputStateTable.hasInputState(sourceId));
        tick.dispatchAll(new DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration());
        assertTrue(inputStateTable.hasInputState(sourceId));
        assertTrue(inputStateTable.getInputState(sourceId).getClass() == KeyInputState.class);
    }

    @Test
    public void shouldAddPointerInputStateWhenRunningActions() throws AppiumException {
        InputStateTable inputStateTable = new InputStateTable();
        Tick tick = new Tick();
        String sourceId = "something2";
        ActionObject actionObject = new ActionObject(sourceId, POINTER, null, 0);
        actionObject.setSubType(POINTER_DOWN);
        actionObject.setButton(0);
        tick.addAction(actionObject);
        assertFalse(inputStateTable.hasInputState(sourceId));
        tick.dispatchAll(new DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration());
        assertTrue(inputStateTable.hasInputState(sourceId));
        assertTrue(inputStateTable.getInputState(sourceId).getClass() == PointerInputState.class);
    }

    @Test
    public void shouldNotAddNoneToState() throws AppiumException {
        InputStateTable inputStateTable = new InputStateTable();
        Tick tick = new Tick();
        String sourceId = "something3";
        ActionObject actionObject = new ActionObject(sourceId, NONE, null, 0);
        tick.addAction(actionObject);
        assertFalse(inputStateTable.hasInputState(sourceId));
        tick.dispatchAll(new DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration());
        assertFalse(inputStateTable.hasInputState(sourceId));
    }

    @Test
    public void shouldDispatchKeyEvents() throws AppiumException {
        InputStateTable inputStateTable = new InputStateTable();
        Tick tick = new Tick();
        String sourceId = "something4";
        KeyInputState keyInputState = new KeyInputState();
        keyInputState.addPressed("g");
        inputStateTable.addInputState(sourceId, keyInputState);
        ActionObject actionObjectOne = new ActionObject(sourceId, KEY, null, 0);
        actionObjectOne.setSubType(KEY_DOWN);
        actionObjectOne.setValue("e");

        ActionObject actionObjectTwo = new ActionObject(sourceId, KEY, null, 0);
        actionObjectTwo.setSubType(KEY_DOWN);
        actionObjectTwo.setValue("f");

        ActionObject actionObjectThree = new ActionObject(sourceId, KEY, null, 0);
        actionObjectThree.setSubType(KEY_UP);
        actionObjectThree.setValue("g");

        ActionObject actionObjectFour = new ActionObject(sourceId, KEY, null, 0);
        actionObjectFour.setSubType(KEY_DOWN);
        actionObjectFour.setValue("\uE008");

        tick.addAction(actionObjectOne);
        tick.addAction(actionObjectTwo);
        tick.addAction(actionObjectThree);
        tick.addAction(actionObjectFour);

        keyInputState = (KeyInputState) inputStateTable.getInputState(sourceId);
        assertFalse(keyInputState.isPressed("e"));
        assertFalse(keyInputState.isPressed("f"));
        assertTrue(keyInputState.isPressed("g"));
        assertFalse(keyInputState.isShift());

        tick.dispatchAll(new DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration());
        assertTrue(keyInputState.isPressed("e"));
        assertTrue(keyInputState.isPressed("f"));
        assertFalse(keyInputState.isPressed("g"));
        assertTrue(keyInputState.isShift());
    }

    @Test
    public void shouldDispatchPointerMoveEvents() throws AppiumException, InterruptedException, ExecutionException {
        final InputStateTable inputStateTable = new InputStateTable();
        KeyInputState keyInputState = new KeyInputState();
        keyInputState.setShift(true);
        inputStateTable.addInputState("keyInputs", keyInputState);
        final Tick tick = new Tick();

        final String sourceId = "something4";
        final String sourceId2 = "something5";
        PointerInputState pointerInputState = new PointerInputState(TOUCH);
        pointerInputState.setType(TOUCH);
        pointerInputState.setX(5L);
        pointerInputState.setY(6L);
        pointerInputState.addPressed(0);
        pointerInputState.addPressed(1);
        inputStateTable.addInputState(sourceId, pointerInputState);
        inputStateTable.addInputState(sourceId2, pointerInputState);

        // Construct pointer move event
        ActionObject actionObjectOne = new ActionObject(sourceId, POINTER, null, 0);
        actionObjectOne.setSubType(POINTER_MOVE);
        actionObjectOne.setPointer(TOUCH);
        actionObjectOne.setX(10L);
        actionObjectOne.setY(20L);
        actionObjectOne.setOrigin(new Origin(VIEWPORT));

        // Construct another pointer move event
        ActionObject actionObjectTwo = new ActionObject(sourceId2, POINTER, null, 0);
        actionObjectTwo.setSubType(POINTER_MOVE);
        actionObjectTwo.setPointer(TOUCH);
        actionObjectTwo.setX(10L);
        actionObjectTwo.setY(20L);
        actionObjectTwo.setOrigin(new Origin(VIEWPORT));

        // Add two pointer move actions to verify that they can run on multiple threads separately
        tick.addAction(actionObjectOne);
        tick.addAction(actionObjectTwo);

        class ExtendedDummyW3CActionAdapter extends DummyW3CActionAdapter {
            @Override
            public void pointerMove(String sourceIdCalled,
                                    PointerType pointerType,
                                    long currentX, long currentY,
                                    long x, long y,
                                    Set<Integer> buttons,
                                    KeyInputState globalKeyInputState) throws AppiumException {
                assertEquals(pointerType, TOUCH);
                assertEquals(currentX, 5L);
                assertEquals(currentY, 6L);
                assertEquals(x, 10L);
                assertEquals(y, 20L);
                assertTrue(buttons.contains(0));
                assertTrue(buttons.contains(1));
                assertTrue(globalKeyInputState.isShift());
            }
        };

        final W3CActionAdapter dummyAdapter = new ExtendedDummyW3CActionAdapter();

        List<Callable<BaseDispatchResult>> callables = tick.dispatchAll(dummyAdapter, inputStateTable, tick.calculateTickDuration());
        Executor executor = Executors.newFixedThreadPool(callables.size());
        CompletionService<BaseDispatchResult> completionService = new ExecutorCompletionService<>(executor);
        for(Callable<BaseDispatchResult> callable:callables) {
            completionService.submit(callable);
        }

        int received = 0;
        while (received < callables.size()) {
            Future<BaseDispatchResult> resultFuture = completionService.take(); //blocks if none available
            resultFuture.get();
            received ++;
        }
    }

    @Test
    public void shouldInvalidCastForSameSourceIdDifferentType() throws AppiumException {
        InputStateTable inputStateTable = new InputStateTable();
        Tick tick = new Tick();
        final String sourceId = "something4";
        PointerInputState pointerInputState = new PointerInputState(TOUCH);
        inputStateTable.addInputState(sourceId, pointerInputState);

        // Depress the shift key
        ActionObject actionObjectTwo = new ActionObject(sourceId, KEY, null, 0);
        actionObjectTwo.setSubType(KEY_DOWN);

        // Construct pointer move event
        ActionObject actionObjectOne = new ActionObject(sourceId, POINTER, null, 0);
        actionObjectOne.setSubType(POINTER_DOWN);

        tick.addAction(actionObjectOne);
        tick.addAction(actionObjectTwo);

        try {
            tick.dispatchAll(new DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration());
        } catch (InvalidArgumentException e) {
            assertTrue(e.getMessage().contains("Attempted to apply action of type"));
            return;
        }

        fail("Should have called 'InvalidArgumentException'");
    }
}

File: espresso-server/library/src/test/java/io/appium/espressoserver/test/helpers/w3c/TickTest.kt
package io.appium.espressoserver.test.helpers.w3c

import org.junit.Test
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorCompletionService
import java.util.concurrent.Executors

import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
import io.appium.espressoserver.lib.helpers.w3c.adapter.DummyW3CActionAdapter
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.BaseDispatchResult
import io.appium.espressoserver.lib.helpers.w3c.models.ActionObject
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType
import io.appium.espressoserver.lib.helpers.w3c.models.Origin
import io.appium.espressoserver.lib.helpers.w3c.models.Tick
import io.appium.espressoserver.lib.helpers.w3c.state.InputStateTable
import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState
import io.appium.espressoserver.lib.helpers.w3c.state.PointerInputState

import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.KEY_DOWN
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.KEY_UP
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.PAUSE
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.POINTER_DOWN
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.POINTER_MOVE
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.KEY
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.NONE
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.POINTER
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType.TOUCH
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail

class TickTest {

    @Test
    fun `should calculate max duration as zero if no durations`() {
        val tick = Tick()
        tick.addAction(ActionObject())
        assertFloatEquals(tick.calculateTickDuration(), 0f)
    }

    @Test
    fun `should calculate the max duration`() {

        val valueOne = arrayOf(10F, 30F, 0F)
        val valueTwo = arrayOf(20F, 10F, 1F)
        val expectedMax = arrayOf(20F, 30F, 1F)

        for (i in valueOne.indices) {
            val tick = Tick()
            val actionObjectOne = ActionObject()
            actionObjectOne.type = NONE
            actionObjectOne.subType = PAUSE
            actionObjectOne.duration = valueOne[i]

            val actionObjectTwo = ActionObject()
            actionObjectTwo.type = POINTER
            actionObjectTwo.subType = POINTER_MOVE
            actionObjectTwo.duration = valueTwo[i]

            val actionObjectThree = ActionObject()
            actionObjectTwo.type = POINTER
            actionObjectTwo.subType = POINTER_MOVE

            tick.addAction(actionObjectOne)
            tick.addAction(actionObjectTwo)
            tick.addAction(actionObjectThree)

            assertFloatEquals(tick.calculateTickDuration(), expectedMax[i])
        }
    }

    @Test
    @Throws(AppiumException::class)
    fun `should add key to key input state when running key actions`() {
        val inputStateTable = InputStateTable()
        val tick = Tick()
        val sourceId = "something1"
        val actionObject = ActionObject(sourceId, KEY, null, 0)
        actionObject.subType = KEY_DOWN
        actionObject.value = "F"
        tick.addAction(actionObject)
        assertFalse(inputStateTable.hasInputState(sourceId))
        tick.dispatchAll(DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration())
        assertTrue(inputStateTable.getInputState(sourceId) is KeyInputState)
    }

    @Test
    @Throws(AppiumException::class)
    fun `should add pointer input state when running pointer actions`() {
        val inputStateTable = InputStateTable()
        val tick = Tick()
        val sourceId = "something2"
        val actionObject = ActionObject(sourceId, POINTER, null, 0)
        actionObject.subType = POINTER_DOWN
        actionObject.button = 0
        tick.addAction(actionObject)
        assertFalse(inputStateTable.hasInputState(sourceId))
        tick.dispatchAll(DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration())
        assertTrue(inputStateTable.getInputState(sourceId) is PointerInputState)
    }

    @Test
    @Throws(AppiumException::class)
    fun `should not affect input state when NONE event is dispatched`() {
        val inputStateTable = InputStateTable()
        val tick = Tick()
        val sourceId = "something3"
        val actionObject = ActionObject(sourceId, NONE, null, 0)
        tick.addAction(actionObject)
        assertFalse(inputStateTable.hasInputState(sourceId))
        tick.dispatchAll(DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration())
        assertFalse(inputStateTable.hasInputState(sourceId))
    }

    @Test
    @Throws(AppiumException::class)
    fun `should dispatch a sequence of key events`() {
        val inputStateTable = InputStateTable()
        val tick = Tick()
        val sourceId = "something4"
        var keyInputState = KeyInputState()
        keyInputState.addPressed("g")
        inputStateTable.addInputState(sourceId, keyInputState)
        val actionObjectOne = ActionObject(sourceId, KEY, null, 0)
        actionObjectOne.subType = KEY_DOWN
        actionObjectOne.value = "e"

        val actionObjectTwo = ActionObject(sourceId, KEY, null, 0)
        actionObjectTwo.subType = KEY_DOWN
        actionObjectTwo.value = "f"

        val actionObjectThree = ActionObject(sourceId, KEY, null, 0)
        actionObjectThree.subType = KEY_UP
        actionObjectThree.value = "g"

        val actionObjectFour = ActionObject(sourceId, KEY, null, 0)
        actionObjectFour.subType = KEY_DOWN
        actionObjectFour.value = "\uE008"

        tick.addAction(actionObjectOne)
        tick.addAction(actionObjectTwo)
        tick.addAction(actionObjectThree)
        tick.addAction(actionObjectFour)

        keyInputState = inputStateTable.getInputState(sourceId) as KeyInputState
        assertFalse(keyInputState.isPressed("e"))
        assertFalse(keyInputState.isPressed("f"))
        assertTrue(keyInputState.isPressed("g"))
        assertFalse(keyInputState.isShift)

        tick.dispatchAll(DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration())
        assertTrue(keyInputState.isPressed("e"))
        assertTrue(keyInputState.isPressed("f"))
        assertFalse(keyInputState.isPressed("g"))
        assertTrue(keyInputState.isShift)
    }

    @Test
    @Throws(AppiumException::class, InterruptedException::class, ExecutionException::class)
    fun `should dispatch a series of pointer move events`() {
        val inputStateTable = InputStateTable()
        val keyInputState = KeyInputState()
        keyInputState.isShift = true
        inputStateTable.addInputState("keyInputs", keyInputState)
        val tick = Tick()

        val sourceId = "something4"
        val sourceId2 = "something5"
        val pointerInputState = PointerInputState(TOUCH)
        pointerInputState.type = TOUCH
        pointerInputState.x = 5f
        pointerInputState.y = 6f
        pointerInputState.addPressed(0)
        pointerInputState.addPressed(1)
        inputStateTable.addInputState(sourceId, pointerInputState)
        inputStateTable.addInputState(sourceId2, pointerInputState)

        // Construct pointer move event
        val actionObjectOne = ActionObject(sourceId, POINTER, null, 0)
        actionObjectOne.subType = POINTER_MOVE
        actionObjectOne.pointer = TOUCH
        actionObjectOne.x = 10.0f
        actionObjectOne.y = 20.0f
        actionObjectOne.origin = Origin(Origin.VIEWPORT)

        // Construct another pointer move event
        val actionObjectTwo = ActionObject(sourceId2, POINTER, null, 0)
        actionObjectTwo.subType = POINTER_MOVE
        actionObjectTwo.pointer = TOUCH
        actionObjectTwo.x = 10.0f
        actionObjectTwo.y = 20.0f
        actionObjectTwo.origin = Origin(Origin.VIEWPORT)

        // Add two pointer move actions to verify that they can run on multiple threads separately
        tick.addAction(actionObjectOne)
        tick.addAction(actionObjectTwo)

        class ExtendedDummyW3CActionAdapter : DummyW3CActionAdapter() {
            @Suppress("UNUSED_PARAMETER", "unused")
            fun pointerMove(sourceId: String,
                            pointerType: PointerType,
                            currentX: Float?, currentY: Float?,
                            x: Float?, y: Float?,
                            buttons: Set<Int>,
                            globalKeyInputState: KeyInputState) {
                assertEquals(pointerType, TOUCH)
                assertFloatEquals(currentX!!, 5f)
                assertFloatEquals(currentY!!, 6f)
                assertFloatEquals(x!!, 10f)
                assertFloatEquals(y!!, 20f)
                assertTrue(buttons.contains(0))
                assertTrue(buttons.contains(1))
                assertTrue(globalKeyInputState.isShift)
            }
        }

        val dummyAdapter = ExtendedDummyW3CActionAdapter()

        val callables = tick.dispatchAll(dummyAdapter, inputStateTable, tick.calculateTickDuration())
        val executor = Executors.newFixedThreadPool(callables.size)
        val completionService = ExecutorCompletionService<BaseDispatchResult>(executor)
        for (callable in callables) {
            completionService.submit(callable)
        }

        var received = 0
        while (received < callables.size) {
            val resultFuture = completionService.take() //blocks if none available
            resultFuture.get()
            received++
        }
    }

    @Test
    @Throws(AppiumException::class)
    fun `should throw invalid cast if using same source ID but different type`() {
        val inputStateTable = InputStateTable()
        val tick = Tick()
        val sourceId = "something4"
        val pointerInputState = PointerInputState(TOUCH)
        inputStateTable.addInputState(sourceId, pointerInputState)

        // Depress the shift key
        val actionObjectTwo = ActionObject(sourceId, KEY, null, 0)
        actionObjectTwo.subType = KEY_DOWN

        // Construct pointer move event
        val actionObjectOne = ActionObject(sourceId, POINTER, null, 0)
        actionObjectOne.subType = POINTER_DOWN

        tick.addAction(actionObjectOne)
        tick.addAction(actionObjectTwo)

        try {
            tick.dispatchAll(DummyW3CActionAdapter(), inputStateTable, tick.calculateTickDuration())
        } catch (e: InvalidArgumentException) {
            assertTrue(e.message!!.contains("Attempted to apply action of type"))
            return
        }

        fail("Should have called 'InvalidArgumentException'")
    }
}
